spine-tools/Spine-Toolbox
# PySide6 port and Python 3.11 support

How to test:

- Make a fresh Python 3.8 - 3.10 virtual or conda environment and activate
- Checkout `qt6` branch from spine-toolbox git
- Do `pip install -r requirements.txt`
- Start app with `python -m spinetoolbox`

If you want to test Python 3.11, install it from `python.org`. **Python 3.11 is not available on Conda yet.**

Code changes:
- Update imports to PySide6
- Use QRecursiveMutex in spine_db_manager because QMutex.Recursive is deprecated
- Use | instead of + in combining flags
- Get enum values with .value when needed. For example, Qt.Checked.value instead of Qt.checked
- Update retrieving some constants that are no longer available as instance variables to class variables, such as QToolButton.InstantPopup
- Use SortFilterProxyModel.setFilterRegularExpression() instead of deprecated setFilterRegExp()
- Change deprecated QLayout.setMargin() to QLayout.setContentsMargins()
- Update saving qsettings() to use QCheckBox.checkState().value because QCheckBox.checkState() cannot be cast to int anymore
- QAbstractItemView.itemDelegate(index) does not exist anymore. Use QAbstractItemView.itemDelegateForIndex() instead.
- Update paint() in project_item_icon.py
- Update pyside2 version check to pyside6 at startup
- Fix Zooming in Design View
- Fix Specification dropdown menu colors in Tool, Importer, and Exporter Properties (**Broken in master branch**)
- Change all exec_() to exec()
- Lots of enum references changed to pyside6 recommendede style, e.g. Qt.CheckState.Checked instead of Qt.Checked
- Fixed SearchBarEditor placement for AlternativeScenarioDelegate and ToolFeatureDelegate (the placement of the little dropdown list with `yes` and `no` when double-clicking the `active: yes` item in Alternative/scenario tree and Tool/feature tree) 
- Updated Export to PDF feature to Qt6 in Spine Db Editor
- Fixed warnings when running tests
- QMouseEvent.pos() is deprecated in pyside6. Replaced with QMouseEvent.position().toPoint() . Note that QWidgets, QEvents, and QGraphicsItems still have a .pos() method or property.

Other changes:
- Update build_ui scripts to use pyside6 tools
- Rebuilt the UI and updated some deprecated enums in .ui files
- Require pyside6 > 6.4
- Install spine packages from compatible branches when running tests
- Fixed tests for Python 3.11 and pyside6
- Removed Python 3.7 support because matplotlib version that is compatible with PySide6 requires Python 3.8 or greater

Fixes #1896

## Checklist before merging
- [x] Unit tests have been added/updated accordingly
- [x] Code has been formatted by black
